llmfoundry/models/hf/hf_base.py

```diff
@@ -331,53 +334,111 @@ def build_inner_model(
 
         dist.barrier()
 
-        # initialize the model on the correct device
-        if resolved_init_device == 'cpu':
-            if pretrained:
-                model = auto_model_cls.from_pretrained(
-                    pretrained_model_name_or_path,
-                    trust_remote_code=trust_remote_code,
-                    use_auth_token=use_auth_token,
-                    load_in_8bit=load_in_8bit,
-                    attn_implementation=requested_attention_implementation,
-                    config=config,
-                )
+        def download_thread_target(queue: Optional[queue.Queue]):
+            """Thread target to wait for the model to be downloaded."""
+            device = get_device(None)
+
+            status = 0
+
+            done_tensor = device.tensor_to_device(
+                torch.tensor(
+                    [status],
+                    dtype=torch.int,
+                ),
+            )
+
+            dist.broadcast(done_tensor, src=0)
+            all_done = done_tensor.item()
+
+            while not all_done:
+                # On rank 0, queue exists, and the presence of an item indicates completion
+                # On all other ranks, the queue object does not exist
+                if queue is not None and not queue.empty():
+                    status = 1
+                done_tensor.fill_(status)
+                dist.broadcast(done_tensor, src=0)
+                all_done = done_tensor.item()
+                time.sleep(1)
+
+        # Create a thread to busy wait for the download to be complete
+        # This approach is used because NCCL process group does not support updating the timeout
+        # and we don't want to require a super long distributed timeout for all operations
+        download_thread = None
+        download_status_queue = None
+        if dist.get_world_size() > 1:
+            download_status_queue = queue.Queue() if dist.get_global_rank(
+            ) == 0 else None
+
+            # Create a thread to busy wait for download to be complete
+            download_thread = threading.Thread(
```

Comment:
nit: download_thread makes it sound like we're downloading something on the thread itself, but we're just waiting. maybe `wait_for_done` or `wait_for_download`

Reply:
done